simdpp/algorithm/transform_reduce.h

```diff
@@ -0,0 +1,150 @@
+/*  Copyright (C) 2018  Povilas Kanapickas <povilas@radix.lt>
+
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at
+    http://www.boost.org/LICENSE_1_0.txt)
+*/
+
+#ifndef LIBSIMDPP_SIMDPP_ALGORITHM_TRANSFORM_REDUCE_H 
+#define LIBSIMDPP_SIMDPP_ALGORITHM_TRANSFORM_REDUCE_H 
+
+#ifndef LIBSIMDPP_SIMD_H
+#error "This file must be included through simd.h"
+#endif
+
+#include <iterator> 
+#include <simdpp/detail/align.h>
+#include <simdpp/types/traits.h>
+#include <simdpp/detail/insn/load_u.h>
+#include <simdpp/detail/insn/load.h>
+#include <simdpp/algorithm/helper_input_range.h>
+
+namespace simdpp {
+   namespace SIMDPP_ARCH_NAMESPACE {
+      template<typename T,typename BinaryOp, typename UnaryOp>
+         T transform_reduce( const T* first, const T* last, T init, BinaryOp binary_op,UnaryOp unary_op)
+      {
+#ifndef NDEBUG //precondition debug mode           
+         if (!first)
+            throw std::runtime_error("transform_reduce - null ptr first.");
+         if (!last)
+            throw std::runtime_error("transform_reduce - null ptr last.");
+#endif
+         using simd_type_T = typename typetraits<T>::simd_type;
+
+         //Define loop counter
+         const auto simd_size = simd_type_T::base_length;
+         const auto size = std::distance(first, last);
+         //note enforce that input is aligned when we start the main simd loop
+         const auto range = helper_input_range(first, last);
+         const auto size_prologue_loop = range.first;
+         const auto size_simd_loop = range.second;
+
+         auto i = 0u;
+         //---prologue
+         for (; i < size_prologue_loop; ++i)
+         {
+            init = binary_op(init,unary_op(*first++));
```

Comment:
I think it makes sense to do possibly overlapping SIMD operations for the prologue and epilogue and then mask out the extra lanes.

Reply:
Sorry i don't understand this point. Could you make an exemple?